packages/remix-dev/compiler/plugins/cssModulesPlugin.ts

```diff
@@ -0,0 +1,387 @@
+  let parcelExports = res.exports || {};
+
+  // sort() to keep order consistent in different builds
+  for (let originClass of Object.keys(parcelExports).sort()) {
+    let props = parcelExports[originClass];
+    let patchedClass = props.name;
+    let prefixedClassName = getPrefixedClassName(classPrefix, patchedClass);
+    json[originClass] = props.composes.length
+      ? getComposedClassNames(classPrefix, prefixedClassName, props.composes)
+      : prefixedClassName;
+    cssModulesContent = cssModulesContent.replace(
+      new RegExp(`\\.${patchedClass}`, "g"),
+      "." + prefixedClassName
+    );
+  }
+
+  let cssWithSourceMap = cssModulesContent;
+  if (res.map) {
+    // TODO: Sourcemaps aren't working as expected because we are inlining the
+    // map at the end of each module. We need to merge them into a single
+    // inline sourcemap at the end of the build. We can probably use something
+    // like https://www.npmjs.com/package/merge-source-maps
+    // cssWithSourceMap += `\n/*# sourceMappingURL=data:application/json;base64,${res.map.toString(
+    //   "base64"
+    // )} */`;
+  }
+
+  return {
+    css: cssWithSourceMap,
+    source: source.toString(),
+    json,
+    moduleExports: parcelExports,
+    dependencies: res.dependencies || [],
+  };
+}
+
+function getResolvedFilePath(
+  config: RemixConfig,
+  args: { path: string; resolveDir: string }
+) {
+  // TODO: Ideally we should deal with the "~/" higher up in the build process
+  // if possible. Also ... what if the user changes this alias in their
+  // tsconfig? Do we support that?
+  return args.path.startsWith("~/")
+    ? path.resolve(config.appDirectory, args.path.replace(/^~\//, ""))
+    : path.resolve(args.resolveDir, args.path);
+}
+
+export function getCssModulesFileReferences(
+  config: RemixConfig,
+  css: string
+): [globalStylesheetFilePath: string, globalStylesheetFileUrl: string] {
+  let hash = getHash(css).slice(0, 8).toUpperCase();
+  let filePath = path.resolve(
+    config.assetsBuildDirectory,
+    "_assets",
+    `__css-modules-${hash}.css`
+  );
+  let fileUrl = resolveUrl(config, filePath);
+  return [filePath, fileUrl];
+}
+
+/**
+ * When a user composes classnames in CSS modules, the value returned for the
+ * JSON map is a concat'd version of all the classname strings composed. Note
+ * that the user may compose classnames referenced in other CSS module files,
+ * but that will require us to juggle dependencies which is a little tricky with
+ * the current build since we don't know for sure if that dependency has been
+ * processed yet. Coming back to this.
+ */
+function getComposedClassNames(
+  prefix: string,
+  name: string,
+  composes: CssModuleReference[]
+) {
+  return composes.reduce((className, composed) => {
+    // skip dependencies for now
+    if (composed.type === "dependency") {
+      console.log({ composed });
+      return className;
+    }
+    return className + " " + getPrefixedClassName(prefix, composed.name);
```

Comment:
FYI, you will need to do this recursively since locals can compose other locals. For example:

```css
.foo {
  color: red;
}

.bar {
  composes: foo;
}

.baz {
  composes: bar;
}
```

When using `baz`, the class names for `bar` and `foo` should also be applied.

It may be helpful to look at the implementation in Parcel here: https://github.com/parcel-bundler/parcel/blob/50893a7ec2b1490e6d0a2f5e1f9ab6a8fc57a113/packages/transformers/css/src/CSSTransformer.js#L153-L187